src/common/com/intellij/plugins/haxe/ide/folding/HaxeFoldingBuilder.java

```diff
@@ -22,37 +23,138 @@
+    final IElementType elementType = node.getElementType();
+
+    FoldingDescriptor descriptor = null;
+    if (isImportOrUsingStatement(elementType) && isFirstImportStatement(node)) {
+      descriptor = buildImportsFolding(node);
+    } else if (isCodeBlock(elementType)) {
+      descriptor = buildCodeBlockFolding(node);
+    } else if (isBodyBlock(elementType)) {
+      descriptor = buildBodyBlockFolding(node);
+    }
+
+    if (descriptor != null) {
+      list.add(descriptor);
     }
+
     for (ASTNode child : node.getChildren(null)) {
       buildFolding(child, list);
     }
   }
 
+  private static boolean isCodeBlock(IElementType elementType) {
+    return elementType == BLOCK_STATEMENT;
+  }
+
+  private static boolean isBodyBlock(IElementType elementType) {
+    return BODY_TYPES.contains(elementType);
+  }
+
+  private static FoldingDescriptor buildCodeBlockFolding(@NotNull ASTNode node) {
+    final ASTNode openBrace = node.getFirstChildNode();
+    final ASTNode closeBrace = node.getLastChildNode();
+
+    return buildBlockFolding(node, openBrace, closeBrace);
+  }
+
+  private static FoldingDescriptor buildBodyBlockFolding(@NotNull ASTNode node) {
+    final ASTNode openBrace = UsefulPsiTreeUtil.getPrevSiblingSkipWhiteSpacesAndComments(node);
+    final ASTNode closeBrace = UsefulPsiTreeUtil.getNextSiblingSkipWhiteSpacesAndComments(node);
+
+    return buildBlockFolding(node, openBrace, closeBrace);
+  }
+
+  private static FoldingDescriptor buildBlockFolding(@NotNull ASTNode node, ASTNode openBrace, ASTNode closeBrace) {
+    TextRange textRange;
+    if (openBrace != null && closeBrace != null && openBrace.getElementType() == PLCURLY && closeBrace.getElementType() == PRCURLY) {
+      textRange = new TextRange(openBrace.getTextRange().getEndOffset(), closeBrace.getStartOffset());
+    } else {
+      textRange = node.getTextRange();
+    }
+
+    if (textRange.getLength() > PLACEHOLDER_TEXT.length()) {
+      return new FoldingDescriptor(node, textRange);
+    }
+
+    return null;
+  }
+
+  private static FoldingDescriptor buildImportsFolding(@NotNull ASTNode node) {
+    final ASTNode lastImportNode = findLastImportNode(node);
+    if (!node.equals(lastImportNode)) {
+      return new FoldingDescriptor(node, buildImportsFoldingTextRange(node, lastImportNode));
+    }
+    return null;
+  }
+
+  private static TextRange buildImportsFoldingTextRange(ASTNode firstNode, ASTNode lastNode) {
+    ASTNode nodeStartFrom = UsefulPsiTreeUtil.getNextSiblingSkipWhiteSpacesAndComments(firstNode.getFirstChildNode());
+    if (nodeStartFrom == null) {
+      nodeStartFrom = firstNode;
+    }
+    return new TextRange(nodeStartFrom.getStartOffset(), lastNode.getTextRange().getEndOffset());
+  }
+
+  private static boolean isFirstImportStatement(@NotNull ASTNode node) {
+    ASTNode prevNode = node.getTreePrev();
+    while (prevNode != null && prevNode.getElementType() == WHITE_SPACE) {
```

Comment:
This fails if there are comments or conditional processing elements in the import list (which there frequently are).   You should rather check against the WHITESPACES and COMMENTS token sets.  So, use UsefulPsiTreeUtil.getPrevSiblingSkippingWhitespaces()?